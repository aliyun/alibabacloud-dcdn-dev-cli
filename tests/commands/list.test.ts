import { describe, expect, it, vi } from 'vitest';
import {
  displayRoutineList,
  handleList
} from '../../src/commands/routine/list.js';
import {
  EdgeFunctionItem,
  GetRoutineUserInfoRes
} from '../../src/libs/interface.js';
import { mockConsoleMethods } from '../helper/mockConsole.js';
import { ApiService } from '../../src/libs/apiService.js';

describe('displayRoutineList', () => {
  let std = mockConsoleMethods();

  it('should display routine list in a table', () => {
    const versionList = [
      {
        RoutineName: 'routine1',
        CreateTime: '2022-01-01',
        Description: 'SGVsbG8gd29ybGQ=',
        SpecName: 'spec1'
      },
      {
        RoutineName: 'routine2',
        CreateTime: '2022-01-02',
        Description: 'VGhpcyBpcyBhIHN0cmluZw==',
        SpecName: 'spec2'
      }
    ];

    displayRoutineList(versionList);

    expect(std.out).toMatchSnapshot();
  });

  it('should display empty table if versionList is empty', () => {
    const versionList: EdgeFunctionItem[] = [];

    displayRoutineList(versionList);
    expect(std.out).toMatchSnapshot();
  });

  it('should display a routine list by calling getRoutineUserInfo', async () => {
    const mockRes: GetRoutineUserInfoRes = {
      Routines: [
        {
          RoutineName: 'routine1',
          Description: 'SGVsbG8gd29ybGQ=',
          CreateTime: '2022-01-01',
          SpecName: 'spec1'
        },
        {
          RoutineName: 'routine2',
          Description: 'VGhpcyBpcyBhIHN0cmluZw==',
          CreateTime: '2022-01-02',
          SpecName: 'spec2'
        }
      ],
      Subdomains: ['subdomain1', 'subdomain2']
    };
    vi.mocked(
      (await ApiService.getInstance()).getRoutineUserInfo
    ).mockResolvedValue(mockRes);
    await handleList({
      _: [],
      $0: ''
    });
    expect(std.out).toMatchSnapshot();
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('Name'));
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('Created'));
    expect(std.out).toHaveBeenCalledWith(
      expect.stringContaining('Description')
    );
    expect(std.out).toHaveBeenCalledWith(
      expect.stringContaining('Specification')
    );

    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('routine1'));
    expect(std.out).toHaveBeenCalledWith(
      expect.stringContaining('Hello world')
    );
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('2022/1/1'));
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('spec1'));

    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('routine2'));
    expect(std.out).toHaveBeenCalledWith(
      expect.stringContaining('This is a string')
    );
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('2022/1/2'));
    expect(std.out).toHaveBeenCalledWith(expect.stringContaining('spec2'));
  });
});
