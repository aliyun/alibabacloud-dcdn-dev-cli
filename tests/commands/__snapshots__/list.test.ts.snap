// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`displayRoutineList > should display a routine list by calling getRoutineUserInfo 1`] = `
[MockFunction log] {
  "calls": [
    [
      "[1m[100m📃 List all of routine:[49m[22m",
    ],
    [
      "[90m┌──────────[39m[90m┬───────────────────[39m[90m┬──────────────────┐[39m
[90m│[39m[31m Name     [39m[90m│[39m[31m Created           [39m[90m│[39m[31m Description      [39m[90m│[39m
[90m├──────────[39m[90m┼───────────────────[39m[90m┼──────────────────┤[39m
[90m│[39m routine1 [90m│[39m 2022/1/1 08:00:00 [90m│[39m Hello world      [90m│[39m
[90m├──────────[39m[90m┼───────────────────[39m[90m┼──────────────────┤[39m
[90m│[39m routine2 [90m│[39m 2022/1/2 08:00:00 [90m│[39m This is a string [90m│[39m
[90m└──────────[39m[90m┴───────────────────[39m[90m┴──────────────────┘[39m",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`displayRoutineList > should display empty table if versionList is empty 1`] = `
[MockFunction log] {
  "calls": [
    [
      "[90m┌──────[39m[90m┬─────────[39m[90m┬─────────────┐[39m
[90m│[39m[31m Name [39m[90m│[39m[31m Created [39m[90m│[39m[31m Description [39m[90m│[39m
[90m└──────[39m[90m┴─────────[39m[90m┴─────────────┘[39m",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`displayRoutineList > should display routine list in a table 1`] = `
[MockFunction log] {
  "calls": [
    [
      "[90m┌──────────[39m[90m┬───────────────────[39m[90m┬──────────────────┐[39m
[90m│[39m[31m Name     [39m[90m│[39m[31m Created           [39m[90m│[39m[31m Description      [39m[90m│[39m
[90m├──────────[39m[90m┼───────────────────[39m[90m┼──────────────────┤[39m
[90m│[39m routine1 [90m│[39m 2022/1/1 08:00:00 [90m│[39m Hello world      [90m│[39m
[90m├──────────[39m[90m┼───────────────────[39m[90m┼──────────────────┤[39m
[90m│[39m routine2 [90m│[39m 2022/1/2 08:00:00 [90m│[39m This is a string [90m│[39m
[90m└──────────[39m[90m┴───────────────────[39m[90m┴──────────────────┘[39m",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
